src/headless-cms/apollo/apolloClient.ts

```diff
@@ -0,0 +1,230 @@
+ */
+function createCmsApolloClient(): ApolloClient<NormalizedCacheObject> {
+  // Rewrite the URLs coming from events API to route them internally.
+  const transformInternalURLs = new ApolloLink((operation, forward) => {
+    return forward(operation).map((response) => {
+      response.data = response.data
+        ? rewriteInternalURLs(response.data)
+        : response.data;
+      return response;
+    });
+  });
+
+  const errorLink = onError(({ graphQLErrors, networkError, operation }) => {
+    if (graphQLErrors) {
+      graphQLErrors.forEach((error) => {
+        const { message, locations, path } = error;
+        const errorMessage = `[GraphQL error]: ${JSON.stringify({
+          OperationName: operation.operationName,
+          Message: message,
+          Location: locations,
+          Path: path,
+          // Response: response,
+        })}`;
+        // eslint-disable-next-line no-console
+        console.error(errorMessage);
+        Sentry.captureMessage(errorMessage);
+      });
+    }
+
+    if (networkError) {
+      // eslint-disable-next-line no-console
+      console.error(
+        `[GraphQL networkError]: ${JSON.stringify({
+          Operation: operation.operationName,
+          NetworkError: networkError,
+        })}`
+      );
+      Sentry.captureMessage('Graphql Network error');
+    }
+  });
+
+  const httpLink = new HttpLink({
+    uri: AppConfig.cmsGraphqlEndpoint,
+    fetch,
+  });
+
+  const cache = createCMSApolloCache();
+
+  const apolloCachePersistor = new TimedApolloCachePersistor(cache, {
+    logLevel: PersistorLogLevel.WARN,
+    persistedCacheTimeToLiveMs: AppConfig.apolloPersistedCacheTimeToLiveMs,
+  });
+
+  // The cache is restored from localStorage if it exists and has not expired.
+  // If the cache has expired, it is purged and the cache is restored from the network.
+  // This is done to avoid flashing on the UI (menus, languages, etc).
+  //
+  // NOTE: The Cache Persistor `purge` and `restore` are async functions, so __we should await them__
+  // to avoid racing conditions: __It might be that the cache is not restored when the first query is made.__
+  // In these cases, the cache will be restored in the background, but a new query will be made to the network,
+  // since we cannot await the async function to finish in a sync function and don't have the cache yet,
+  // A React-Router and it's navigation components and tools helps to hide this issue, since the user can navigate
+  // without rerendering the whole app again and then the Apollo cache is not cleared during the route changes.
+  (async () => {
+    if (apolloCachePersistor.hasPersistedCacheExpired()) {
+      // eslint-disable-next-line no-console
+      console.info('Persisted cache has expired.');
+      await apolloCachePersistor.purge();
+    } else {
+      await apolloCachePersistor.restore();
+      // eslint-disable-next-line no-console
+      console.info('Persisted cache has been restored.');
+    }
+  })();
```

Comment:
Note this racing condition. Sometimes the cache is restored fast enough, but sometimes not. However, the only issue should be that a needless CMS query is been made. I cannot change the Apolloclient to be a promise, because we cannot await for it everywhere were it's needed. It should be a static variable that is always available.